src/core/inductive.ml

```diff
@@ -0,0 +1,260 @@
+  fun pos ind_sym codom ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+module P  =
+  struct
+
+    (** [iden s] creates an identifier without position thanks to the string
+        [s]. *)
+    let iden : string -> p_term = fun s ->
+      Pos.none (P_Iden(Pos.none ([], s), true))
+
+    (** [patt s ts] creates a pattern without position thanks to the string
+        [s] and the terms [ts]. *)
+    let patt : string -> p_term array option -> p_term = fun s ts ->
+      Pos.none (P_Patt (Some (Pos.none s), ts))
+
+    (** [patt0 s] creates a pattern without position thanks to the string
+        [s]. *)
+    let patt0 : string -> p_term = fun s -> patt s None
+
+    (** [appl t1 t2] creates an application of [t1] to [t2] without
+        position. *)
+    let appl : p_term -> p_term -> p_term = fun t1 t2 ->
+      Pos.none (P_Appl(t1, t2))
+
+    (** [nb_wild head i] creates a p_term which has [i] wildcard(s) after the
+        head [head]. *)
+    let nb_wild : p_term -> int -> p_term = fun head i ->
+      let rec aux acc i =
+        if i = 0 then acc else aux (appl acc (Pos.none P_Wild)) (i-1)
+      in
+      aux head i
+  end
+
+(** [fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
+    cons_sym f_rec f_not_rec acc] generates some value iteratively by going
+    through the structure of [cons_sym.sym_type]. The position [pos] is the
+    position of the command inductive where the constructors were defined.
+    On the codomain, the function [codom] is called on the arguments to which
+    [ind_sym] is applied and the variables of the products (in reverse order).
+    In case of a product, the function [get_name] splits the product and
+    naming the variable if it is not the case. In case of recursive occurren-
+    ces, the function [build_rec_hyp] builds the recursive hypothesis associa-
+    ted, and then the function [domrec] is applied to conclude the building.
+    Otherwise, the function [dom] is called. If you would like to store a tem-
+    porary result, the initial value is [acc], and you can change this value
+    in the recursive case with the function [f_rec], and on the other case
+    with the function [f_not_rec]. *)
+let fold_cons_typ :
+      popt ->
+      (term list -> 'b list -> 'a -> 'c) ->
+      ( (term, term) Bindlib.binder -> int -> 'b * term ) ->
+      (term list -> 'b -> 'a) ->
+      (term -> 'b -> 'a -> 'c -> 'c) ->
+      (term -> 'b -> 'c -> 'c) ->
+      sym -> sym ->
+      ('a -> 'b -> 'a -> 'a) ->
+      ('a -> 'b -> 'a) ->
+      'a ->
+      'c =
+  fun pos codom get_name build_rec_hyp domrec dom ind_sym
+      cons_sym f_rec f_not_rec acc->
```

Comment:
There are many arguments so that we don't know the type of each one. I suggest to use the following notation instead:
``` let fold_cons_typ (pos : popt) (codom : ...) ... : 'c =